modules/bm_sim/include/bm_sim/match_units.h

```diff
@@ -377,94 +372,29 @@ class MatchUnitAbstract : public MatchUnitAbstract_ {
+template <typename K, typename V>
+class MatchUnitGeneric : public MatchUnitAbstract<V> {
  public:
   typedef typename MatchUnitAbstract<V>::MatchUnitLookup MatchUnitLookup;
-
- public:
-  MatchUnitExact(size_t size, const MatchKeyBuilder &match_key_builder)
-    : MatchUnitAbstract<V>(size, match_key_builder),
-      entries(size) {
-    entries_map.reserve(size);
-  }
-
- private:
-  // TODO(antonin): have all Entry structs inherit from a common base?
   struct Entry {
-    Entry() { }
-
-    Entry(ByteContainer key, V value, uint32_t version)
-      : key(std::move(key)), value(std::move(value)), version(version) { }
-
-    ByteContainer key{};
-    V value{};
-    uint32_t version{0};
-
-    static constexpr MatchUnitType mut = MatchUnitType::EXACT;
+    Entry() {}
+    Entry(K key, V value)
+      : key(std::move(key)), value(std::move(value)) {}
+    K key;
+    V value;
   };
 
- private:
-  MatchErrorCode add_entry_(const std::vector<MatchKeyParam> &match_key,
-                            V value,  // by value for possible std::move
-                            entry_handle_t *handle,
-                            int priority) override;
-
-  MatchErrorCode delete_entry_(entry_handle_t handle) override;
-
-  MatchErrorCode modify_entry_(entry_handle_t handle, V value) override;
-
-  MatchErrorCode get_value_(entry_handle_t handle, const V **value) override;
-
-  MatchErrorCode get_entry_(entry_handle_t handle,
-                            std::vector<MatchKeyParam> *match_key,
-                            const V **value, int *priority) const override;
-
-  MatchErrorCode dump_match_entry_(std::ostream *out,
-                                   entry_handle_t handle) const override;
-
-  void dump_(std::ostream *stream) const override;
-
-  void reset_state_() override;
-
-  MatchUnitLookup lookup_key(const ByteContainer &key) const override;
-
- private:
-  std::vector<Entry> entries{};
-  std::unordered_map<ByteContainer, entry_handle_t, ByteContainerKeyHash>
-  entries_map{};
-};
-
-template <typename V>
-class MatchUnitLPM : public MatchUnitAbstract<V> {
  public:
-  typedef typename MatchUnitAbstract<V>::MatchUnitLookup MatchUnitLookup;
-
- public:
-  MatchUnitLPM(size_t size, const MatchKeyBuilder &match_key_builder)
-    : MatchUnitAbstract<V>(size, match_key_builder),
-      entries(size),
-      entries_trie(this->nbytes_key) { }
-
- private:
-  struct Entry {
-    Entry() { }
-
-    Entry(ByteContainer key, int prefix_length, V value, uint32_t version)
-      : key(std::move(key)), prefix_length(prefix_length),
-        value(std::move(value)), version(version) { }
-
-    ByteContainer key{};
-    int prefix_length{0};
-    V value{};
-    uint32_t version{0};
-
-    static constexpr MatchUnitType mut = MatchUnitType::LPM;
-  };
+  MatchUnitGeneric(size_t size, const MatchKeyBuilder &match_key_builder,
+                   LookupStructureFactory *lookup_factory)
+    : MatchUnitAbstract<V>(size, match_key_builder)
```

Comment:
Like this? (The `{}` on the following line is to make the above line < 80 chars)
```cppp
MatchUnitGeneric(size_t size, const MatchKeyBuilder &match_key_builder,
                   LookupStructureFactory *lookup_factory)
    : MatchUnitAbstract<V>(size, match_key_builder), entries(size),
      lookup_structure(
        LookupStructureFactory::create<K>(lookup_factory,
                                          size,
                                          match_key_builder.get_nbytes_key()))
  {}
```